### PR TITLE
Rule out negative time durations

### PR DESCRIPTION
**Pull request by coventry**
Type `time.Duration` is a type alias of `int64`, so negative durations are possible. Prior to this PR, `models.Duration` was an alias to `time.Duration`, and was used in ways which assume a positive duration, e.g. passing them timeouts (where a negative value can cause a `panic`) and as subtrahends from `time.Time`s to obtain bounds for identifying old records (where a negative value would result in a bound which includes all records.)

This makes `models.Duration` a wrapper around `time.Duration` instead of just a type alias. The only way to create a non-zero `models.Duration` outside the `models` package is via `NewDuration(d time.Duration) (Duration, error)`, which rejects negative `time.Duration`s.

### CODE REVIEW

File: core/services/gas_updater.go [hunk 1]
```diff
@@ -31,8 +31,6 @@ var (
 	)
 )
 
-//go:generate mockery -name GasUpdater  -output ../internal/mocks/ -case=underscore
```
Comment (j16r) [inline core/services/gas_updater.go:34]: Why was this removed?

File: core/services/reaper.go [hunk 2]
```diff
@@ -22,7 +22,8 @@ func NewStoreReaper(store *store.Store) SleeperTask {
 }
 
 func (sr *storeReaper) Work() {
-	offset := time.Now().Add(-sr.config.ReaperExpiration()).Add(-sr.config.SessionTimeout())
+	offset := time.Now().Add(-sr.config.ReaperExpiration().Duration()).
```
Comment (j16r) [inline core/services/reaper.go:25]: Would this be a bit clearer if the two durations were added together than subtracted from time.Now? the two negative durations here feels weird.

File: core/services/reaper_test.go [hunk 3]
```diff
@@ -28,9 +28,10 @@ func TestStoreReaper_ReapSessions(t *testing.T) {
 		wantReap bool
 	}{
 		{"current", time.Now(), false},
-		{"expired", time.Now().Add(-store.Config.SessionTimeout()), false},
-		{"almost stale", time.Now().Add(-store.Config.ReaperExpiration()), false},
-		{"stale", time.Now().Add(-store.Config.ReaperExpiration()).Add(-store.Config.SessionTimeout()), true},
+		{"expired", time.Now().Add(-store.Config.SessionTimeout().Duration()), false},
```
Comment (j16r) [inline core/services/reaper_test.go:31]: These tests look like they mirror the implementation and might not be testing anything. Mind fixing them up?

File: core/store/models/common.go [hunk 4]
```diff
@@ -425,20 +425,41 @@ func (c Cron) String() string {
...
+	return rv
+}
 
 // Duration returns the value as the standard time.Duration value.
 func (d Duration) Duration() time.Duration {
-	return time.Duration(d)
+	return d.d
 }
 
+// Shorter returns true iff d is shorter than od.
```
Comment (j16r) [inline core/store/models/common.go:451]: `iff` is that a typo?

File: core/internal/mocks/caller_subscriber.go [hunk 5]
```diff
@@ -2,12 +2,9 @@
 
 package mocks
 
-import (
-	context "context"
-
-	eth "github.com/smartcontractkit/chainlink/core/eth"
-	mock "github.com/stretchr/testify/mock"
-)
+import context "context"
```
Comment (j16r) [inline core/internal/mocks/caller_subscriber.go:5]: Could you update your generator dependencies and redo these? We keep switching between import formats in mocks because everyone has different versions of those packages. (OR let me know if yours are the latest)

File: core/services/fluxmonitor/flux_monitor.go [hunk 6]
```diff
@@ -250,7 +252,7 @@ func (f pollingDeviationCheckerFactory) New(
 	}
 
 	fetcher, err := newMedianFetcherFromURLs(
-		timeout,
+		timeout.Duration(),
```
Comment (j16r) [inline core/services/fluxmonitor/flux_monitor.go:255]: Shouldn't the `models.Duration` be pushed further down here?

File: core/store/models/common.go [hunk 7]
```diff
@@ -425,20 +425,41 @@ func (c Cron) String() string {
 	return string(c)
 }
 
-// Duration is a time duration.
-type Duration time.Duration
+// Duration is a non-negative time duration.
+type Duration struct{ d time.Duration }
+
+func NewDuration(d time.Duration) (Duration, error) {
```
Comment (j16r) [inline core/store/models/common.go:431]: `New` is typically used to indicate returning a pointer, `Make` is used to denote copies.

**Comment by j16r** [comment]
Can you explain this "and as subtrahends from time.Times to obtain bounds for identifying old records" ?

File: core/services/gas_updater.go (hunk 1 above)
Comment (coventry) [reply core/services/gas_updater.go:34]: It's unused, and its output is not checked in.

File: core/services/reaper.go (hunk 2 above)
Comment (coventry) [reply core/services/reaper.go:25]: > Can you explain this "and as subtrahends from time.Times to obtain bounds for identifying old records" ?

Like in `time.Now().Add(-sr.config.ReaperExpiration().Duration())`.

Comment (coventry) [reply core/services/reaper.go:25]: Oops, don't know why that comment landed there. Deleting.

**Comment by coventry** [comment]
> Can you explain this "and as subtrahends from time.Times to obtain bounds for identifying old records" ?

Like in `time.Now().Add(-sr.config.ReaperExpiration().Duration())`.

File: core/services/reaper.go (hunk 2 above)
Comment (coventry) [reply core/services/reaper.go]: I created a `Durations.Before(time.Time) time.Time` method, so this became `sr.config.ReaperExpiration().Before(sr.config.SessionTimeout().Before(time.Now()))`

File: core/services/reaper_test.go (hunk 3 above)
Comment (coventry) [reply core/services/reaper_test.go:31]: I think these are testing the reaping functionality (i.e., that `DeleteStaleSessions` is doing the right thing.)

File: core/store/models/common.go (hunk 4 above)
Comment (coventry) [reply core/store/models/common.go:456]: Short for "if and only if". I will expand it.

File: core/internal/mocks/caller_subscriber.go (hunk 5 above)
Comment (coventry) [reply core/internal/mocks/caller_subscriber.go:5]: Oops, thanks.

Comment (coventry) [reply core/internal/mocks/caller_subscriber.go:5]: I had `mockery` version `1.0.0` before and after `go get github.com/vektra/mockery/.../`. However, `go generate` still resulted in some differences specifically pertaining to the change to `models.Duration`, which are now the only contents of the git commit tagged as `go generate`.

File: core/services/fluxmonitor/flux_monitor.go (hunk 6 above)
Comment (coventry) [reply core/services/fluxmonitor/flux_monitor.go:255]: That's a nice-to-have, but this is a call to an internal method, and we should really only have to protect `fluxmonitor` at its interfaces.

**Comment by samsondav** [comment]
Was this causing any bugs?

File: core/store/models/common.go [hunk 8]
```diff
@@ -425,20 +425,46 @@ func (c Cron) String() string {
 	return string(c)
 }
 
-// Duration is a time duration.
-type Duration time.Duration
+// Duration is a non-negative time duration.
```
Comment (samsondav) [inline core/store/models/common.go:428]: I cannot believe golang's default allows for negative durations.

What were they thinking? Time travel?

File: core/store/models/common.go [hunk 9]
```diff
@@ -457,10 +483,28 @@ func (d *Duration) UnmarshalJSON(input []byte) error {
...
 	}
-	*d = Duration(v)
+	*d, err = MakeDuration(v)
+	if err != nil {
+		return err
+	}
 	return nil
 }
 
+func (d *Duration) Scan(v interface{}) (err error) {
```
Comment (samsondav) [inline core/store/models/common.go:493]: Postgres actually has a built-in duration type (called `interval`).

We might consider using that instead.

File: core/store/models/job_spec.go [hunk 10]
```diff
@@ -237,6 +237,9 @@ func (i *InitiatorParams) SetDefaultValues(typ string) {
 	if typ == InitiatorFluxMonitor {
 		err := mergo.Merge(i, &FluxMonitorDefaultInitiatorParams)
 		logger.PanicIf(errors.Wrap(err, "type level dependent error covered by tests"))
+		if i.PollingInterval.IsInstant() { // mergo can't handle structs well
```
Comment (samsondav) [inline core/store/models/job_spec.go:240]: What is mergo?

Comment (samsondav) [reply core/store/models/job_spec.go:240]: And why is it relevant? This comment could be clearer

File: core/store/models/common.go (hunk 8 above)
Comment (coventry) [reply core/store/models/common.go:428]: They provide a handy way to compute the time some duration before another time, I guess. I added a `Before` method to do that, though.

File: core/store/models/common.go (hunk 9 above)
Comment (coventry) [reply core/store/models/common.go:493]: I'm reluctant because that increases our dependency on postgres. What would the benefits be?

File: core/store/models/job_spec.go (hunk 10 above)
Comment (coventry) [reply core/store/models/job_spec.go:240]: [Mergo](https://github.com/imdario/mergo) magically merges recursive data structures, replacing empty fields in the destination with the values in the source. The `mergo.Merge` call ought to replace `i.PollingInterval` with the default value, but it doesn't.

`mergo` doesn't accomplish much, and I am inclined to rip it out.

File: core/store/models/common.go (hunk 9 above)
Comment (coventry) [reply core/store/models/common.go:493]: Thanks, I created story #172460686 for that.

File: core/services/fluxmonitor/flux_monitor.go (hunk 6 above)
Comment (j16r) [reply core/services/fluxmonitor/flux_monitor.go:255]: Does "That's a nice to have" just mean "that's out of scope"?

As a rule, I don't entirely agree with `and we should really only have to protect fluxmonitor at its interfaces`, given your PR doesn't really do that - it introduces a new type! And ... flux monitor's interface still accepts negative durations.. maybe some confusion in terminology here.

This may be an appropriate compromise, but there are definitely some ways negative values could sneak through after this PR.

Comment (coventry) [reply core/services/fluxmonitor/flux_monitor.go]: Thanks, you're right that `NewPollingDeviationChecker` is logically part of fluxmonitor's external interface (I ignored it because we don't actually use it externally), and also that it's much better to just use `models.Duration` everywhere. I've removed all instances of `time.Duration` from `core/services/fluxmonitor`. Can I get a re-approval?

**Comment by samsondav** [comment]
@coventry yup all good
